### PR TITLE
Adding new attributes to IssueEvent

### PR DESCRIPTION
**Pull request by allevin**
See Issue #855 

The class [IssueEvent](https://github.com/PyGithub/PyGithub/blob/master/github/IssueEvent.py) is missing a large number of attributes documented in the [API](https://developer.github.com/v3/issues/events/).

This is also commented about in #653 to a degree

27 of the tests 27 known event types have tests.

**Currently Tested using Issue #30**            
- [x] subscribed
- [x] assigned
- [x] referenced               
- [x] closed                   
- [x] labeled                  

**Currently Tested using Issue/PR #538**
- [x] merged
- [x] mentioned
- [x] review_requested

**Currently Tested using Issue/PR #857**
- [x] reopened
- [x] unassigned
- [x] unlabeled
- [x] renamed
- [x] base_ref_changed
- [x] head_ref_deleted 
- [x] head_ref_restored
- [x] milestoned
- [x] demilestoned
- [x] locked
- [x] unlocked
- [x] review_dismissed
- [x] review_request_removed
- [x] marked_as_duplicate
- [x] unmarked_as_duplicate
- [x] added_to_project       
- [x] moved_columns_in_project
- [x] removed_from_project
- [x] converted_note_to_issue - Note: this event is tied into Issue #866 

This PR is now ready to be merged

### CODE REVIEW

**Comment by allevin** [comment]
My testing has failed on Travis, but passes on my local machine.  

It seems from the Travis output that there is a mismatch in the authorization headers.  I am using an auth token to login (I use 2 factor auth), and it records :
``
{'Authorization': 'token private_token_removed', 'User-Agent': 'PyGithub/Python'}
``

However when testing is done in Travis, it seems to be using a username/password auth and is capturing:
``
{'Authorization': 'Basic login_and_password_removed', 'User-Agent': 'PyGithub/Py [truncated]...
``
Since these 2 fields don't match my tests are failing.  

The problem seems to be in the fixAuthorizationHeader function in Framework.py, it looks to change the headers slightly for security, but the live data from travis does not match the recorded data when using a Token.

Am I doing something wrong on my end?

**Comment by allevin** [comment]
@sfdye I need someone with admin privileges to make the changes to this PR described above so I can  write the additional tests for the new supported attributes.  Most of them should be quite easy to implement (label on then off; assign a user then remove, Lock then unlock, etc).  Once I get a few of these I will take another pass at integrating more tests.  I expect this to take a few iterations to before we have all the event accounted for.

**Comment by sfdye** [comment]
Done the first three items. Can you check if it worked for you? If so, I will do the rest.

**Comment by allevin** [comment]
@sfdye Worked perfectly!  Was able to generate the tests for the 4 new events.

**Comment by sfdye** [comment]
All done except how exactly to achieve the following?

marked_as_duplicate - Mark the pr as a duplicate of this issue
unmarked_as_duplicate - Unmark the pr as a duplicate of this issue

**Comment by allevin** [comment]
Duplicate of #855

**Comment by allevin** [comment]
@sfdye Getting closer.  I tried to mark as duplicate, but apparently my attempt went in as a comment.  I suspect I need Admin powers to make it actually happen [see here](https://help.github.com/articles/about-duplicate-issues-and-pull-requests/).

Also, your removal of requested user sfdye did not seem to work (the sfdye is still assigned as a reviewer), you may need to assign someone else and then remove them?

Afthe duplicates and review_request_removed, the only remaining events are related to projects.

**Comment by sfdye** [comment]
Duplicate of #855

**Comment by sfdye** [comment]
How about now?

**Comment by allevin** [comment]
@sfdye Got the marked as duplicate done (now we need to unmark as duplicate).  Also got removed request done.

The remaining events are unmark duplicate and then the project related events described above).

I was having problems getting the latest IssueEvents.  Apparently we had a total of 30+ of them, and the PaginatedList() object was not iterating past the first page of 30.  This had me tripped up for a bit, but basically I realized that this is a feature/issue with the way Github paginates issues events ([See Here](https://developer.github.com/v3/activity/events/)). 

Basically the paginated list will ignore the `per_page` variable of PaginatedList().  The user **must**  execute a PaginatedList.get_page(n) for each page and can only iterate that page.   Then the user must then execute PaginatedList.get_page(n+1) and iterate that page, etc.  PyGithub's PaginatedList() does not handle this automatically (perhaps a future PR?)

I added a documentation comment in the Issues.py the attempts to describe this, but there may be a better way of saying this.

**Comment by sfdye** [comment]
> perhaps a future PR?

Yes, definitely. 

Un-mark duplicate done.

image

I thought I've already triggered the projects related events. Did you NOT get them?

**Comment by allevin** [comment]
@sfdye 

> I thought I've already triggered the projects related events. Did you NOT get them?

Very interesting, I do not see any project related events for this PR (issue) in either the conversation or in the list of events via the code `self.g.get_repo("PyGithub/PyGithub").get_issue(857).get_events()`.  

If you can see it in the conversation, perhaps this is an maintainer/admin level event that can't be seen by regular developers? or perhaps the project was deleted (I don't see it in the projects tab).

Not sure what to do here...

**Comment by sfdye** [comment]
@allevin Found the reason. I've actually created a project at the org level (the screenshot), not at repo level. The org admin has disabled repo level projects, so I need to ask permission to re-enable this. (I don't have permission to do so at the  moment)

**Comment by allevin** [comment]
@sfdye Good to know.  Does this have have any impact to #854?

**Comment by sfdye** [comment]
Should not, since the author of that PR used his own project for the test.

**Comment by sfdye** [comment]
@allevin I just made a public project and linked to this repo. Did you get the remaining events now?

**Comment by allevin** [comment]
@sfdye That did it, and I also caught the event on issue #866.  I have tests written for all documented events (plus an additional event that I stumbled upon).  As far as I am concerned, this PR is finished.  Thanks for your assistance!   Hopefully, this will make it into the next release.

File: github/IssueEvent.py [hunk 1]
```diff
@@ -101,6 +101,105 @@ def url(self):
         self._completeIfNotSet(self._url)
         return self._url.value
 
+    # New Code
```
Comment (sfdye) [inline github/IssueEvent.py]: What's this :)

File: github/Issue.py [hunk 2]
```diff
@@ -387,6 +387,9 @@ def get_events(self):
         """
         :calls: `GET /repos/:owner/:repo/issues/:issue_number/events <http://developer.github.com/v3/issues/events>`_
         :rtype: :class:`github.PaginatedList.PaginatedList` of :class:`github.IssueEvent.IssueEvent`
+        Note: PaginatedList of events does not support per_page option, user must call
```
Comment (sfdye) [inline github/Issue.py]: I am not sure this is the right place to put it

File: github/IssueEvent.py [hunk 3]
```diff
@@ -101,6 +101,105 @@ def url(self):
...
+
+    @property
+    def dismissed_review(self):
+        """
+        :type: dict
+        """
+        self._completeIfNotSet(self._dismissed_review)
+        return self._dismissed_review.value
+
+#    # lock_reason is documented in API, but not provided in response
```
Comment (sfdye) [inline github/IssueEvent.py]: You are right. Let me confirm this with Github support

**Comment by sfdye** [comment]
LGTM. Let's wait what Github staff has to say about the missing attribute.

File: github/IssueEvent.py (hunk 1 above)
Comment (allevin) [reply github/IssueEvent.py]: DOH!  I missed that one ;-(

**Comment by sfdye** [comment]
@allevin Just got back from Github, it looks like there is a hidden custom header `application/vnd.github.sailor-v-preview+json` in order to see this attribute. Wanna try again 😆

**Comment by allevin** [comment]
@sfdye Ok well the changes to IssueEvent and associated calls (Issue.get_events(), Repository.get_issues_event(), and Repository.get_issues_events()) with the custom header seem to work properly.  

However,  the changes to Issue.py and Repository.py wont pass the tests (I assume because of the custom header).  

I tried to record them, but am getting some odd errors...

```
$ python -m github.tests Issue.setUp --record --auth_with_token
GET /user None {'Authorization': 'token xxxxx', 'User-Agent': 'PyGithub/Python'} => 200
GET /repos/allevin/PyGithub None {'Authorization': 'token xxxxx', 'User-Agent': 'PyGithub/Python'} => 200
GET /repos/allevin/PyGithub/issues/28 None {'Authorization': 'token xxxxx', 'User-Agent': 'PyGithub/Python'} => 404
E
======================================================================
ERROR: setUp (github.tests.Issue.Issue)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "github/tests/Issue.py", line 44, in setUp
    self.issue = self.repo.get_issue(28)
  File "github/Repository.py", line 1837, in get_issue
    self.url + "/issues/" + str(number)
  File "github/Requester.py", line 260, in requestJsonAndCheck
    return self.__check(*self.requestJson(verb, url, parameters, headers, input, self.__customConnection(url)))
  File "github/Requester.py", line 271, in __check
    raise self.__createException(status, responseHeaders, output)
UnknownObjectException: 404 {u'documentation_url': u'https://developer.github.com/v3/issues/#get-a-single-issue', u'message': u'Not Found'}

----------------------------------------------------------------------
Ran 1 test in 2.943s

FAILED (errors=1)
```

I tried an absolutely clean clone of PyGithub using the master branch, and the above command `python -m github.tests Issue.setUp --record --auth_with_token` fails exactly the same, so I am not convinced that my code changes are causing the error.  I am unsure what's happening here.  

I am seeing a different failures on recording tests for Repository.py (again using PyGithub/master)

```
[allevin@s957431 PyGithub_test1] (master)*$ python -m github.tests Repository.testGetIssuesEvents --record --auth_with_token
GET /user None {'Authorization': 'token xxxxx', 'User-Agent': 'PyGithub/Python'} => 200
GET /repos/allevin/PyGithub None {'Authorization': 'token xxxxx', 'User-Agent': 'PyGithub/Python'} => 200
GET /repos/allevin/PyGithub/issues/events None {'Authorization': 'token xxxxx', 'User-Agent': 'PyGithub/Python'} => 200
E
======================================================================
ERROR: testGetIssuesEvents (github.tests.Repository.Repository)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "github/tests/Repository.py", line 409, in testGetIssuesEvents
    self.assertListKeyBegin(self.repo.get_issues_events(), lambda e: e.event, ["assigned", "subscribed", "closed", "assigned", "closed"])
  File "github/tests/Framework.py", line 272, in assertListKeyBegin
    realKeys = [key(element) for element in elements[: len(expectedKeys)]]
  File "github/PaginatedList.py", line 91, in __iter__
    yield self.__list[index]
  File "github/PaginatedList.py", line 55, in __getitem__
    self.__fetchToIndex(index)
  File "github/PaginatedList.py", line 73, in __fetchToIndex
    self._grow()
  File "github/PaginatedList.py", line 76, in _grow
    newElements = self._fetchNextPage()
  File "github/PaginatedList.py", line 196, in _fetchNextPage
    headers=self.__headers
  File "github/Requester.py", line 262, in requestJsonAndCheck
    return self.__check(*self.requestJson(verb, url, parameters, headers, input, self.__customConnection(url)))
  File "github/Requester.py", line 319, in requestJson
    return self.__requestEncode(cnx, verb, url, parameters, headers, input, encode)
  File "github/Requester.py", line 372, in __requestEncode
    status, responseHeaders, output = self.__requestRaw(cnx, verb, url, requestHeaders, encoded_input)
  File "github/Requester.py", line 396, in __requestRaw
    response = cnx.getresponse()
  File "github/tests/Framework.py", line 126, in getresponse
    self.__writeLine(str(output))
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2013' in position 28729: ordinal not in range(128)

----------------------------------------------------------------------
Ran 1 test in 2.989s

FAILED (errors=1)
```

File: github/Issue.py [hunk 4]
```diff
@@ -395,7 +395,8 @@ def get_events(self):
             github.IssueEvent.IssueEvent,
             self._requester,
             self.url + "/events",
-            None
+            None,
+            headers={'Accept': 'application/vnd.github.sailor-v-preview+json'}
```
Comment (sfdye) [inline github/Issue.py]: Why do you have to add header here?

Comment (sfdye) [reply github/Issue.py]: BTW, all headers now have a designated place:
https://github.com/PyGithub/PyGithub/blob/master/github/Consts.py

Comment (sfdye) [reply github/Issue.py]: It was introduced in this ae5cdb285e2d41e55b2f6fac56fd4eff239c6309

Comment (allevin) [reply github/Issue.py]: Issue.get_events() returns a IssueEvent object, we need the header here to ensure we get the lock_reason attribute.

**Comment by allevin** [comment]
@sfdye It was a bit of a battle, but I believe I got the custom header to support the `lock_reason` attribute in.  The recorded test files were giving me fits, and I accidently added a few comments to #28  in my thrashing around trying to figure things out (via recording tests/Issue.py).  Sorry about that.

**Comment by sfdye** [comment]
Awesome! We can close this PR now.

**Comment by sfdye** [comment]
@allevin Thank you for all the efforts put in this PR. Feel free to merge in :)

**Comment by allevin** [comment]
@sfdye Thank you for your patience and assistance.  This one was a real effort.  Now I only have 2 more open PR's to sucker you to merge ;-)

> @allevin Thank you for all the efforts put in this PR. Feel free to merge in :)

I don't understand "merge in".  I don't have permissions to merge this PR...
